stablehlo/transforms/StablehloAggressiveSimplification.cpp

```diff
@@ -691,6 +694,121 @@ struct EmptyReduceOpCanon final : OpRewritePattern<mlir::stablehlo::ReduceOp> {
   }
 };
 
+struct UnusedResultReduceOpCanon final
+    : OpRewritePattern<mlir::stablehlo::ReduceOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(mlir::stablehlo::ReduceOp op,
+                                PatternRewriter &rewriter) const override {
+    SmallVector<OpResult, 4> usedResults;
+    llvm::copy_if(op.getResults(), std::back_inserter(usedResults),
+                  [](OpResult result) { return !result.use_empty(); });
+
+    if (usedResults.size() == op.getNumResults())
+      return rewriter.notifyMatchFailure(op, "all operation results have uses");
+
+    const auto pairSize = 2;
+    const auto numOperands = op.getNumOperands();
+    const auto numOperandPairs = numOperands / pairSize;
+
+    Block &reducerBlock = op.getBody().front();
+    auto retOp = cast<mlir::stablehlo::ReturnOp>(reducerBlock.getTerminator());
+
+    SmallVector<unsigned> argToPairMap(numOperands);
+    for (auto i : llvm::seq(0u, numOperandPairs)) {
+      argToPairMap[i] = i;
+      argToPairMap[i + numOperandPairs] = i;
+    }
+
+    SmallVector<Value> workList;
+    auto addToWorkList = [&workList,
+                          retOpRegion = retOp->getParentRegion()](Value v) {
+      // inspect only values from reducerBlock's body
+      if (v.getParentRegion() == retOpRegion) workList.push_back(v);
+    };
+
+    SmallPtrSet<Operation *, 16> usedOps;
+    SmallBitVector usedArgs(numOperands);
+    SmallBitVector usedReturnOperands(numOperandPairs);
+    for (const auto &usedResult : usedResults) {
+      auto resultNo = usedResult.getResultNumber();
+      usedReturnOperands.set(resultNo);
+
+      // Follow the def-use chain starting from return operand to identify
+      // which argument pairs are used to compute it.
+      addToWorkList(retOp.getOperand(resultNo));
+      while (!workList.empty()) {
+        auto definition = workList.pop_back_val();
+        if (auto blockArg = definition.dyn_cast<BlockArgument>()) {
+          const auto pairNo = argToPairMap[blockArg.getArgNumber()];
+          // using one argument implies using the whole argument pair
+          usedArgs.set(pairNo);
+          usedArgs.set(pairNo + numOperandPairs);
+        } else if (auto *defOp = definition.getDefiningOp()) {
+          usedOps.insert(defOp);
+          for (const auto &operand : defOp->getOperands())
+            addToWorkList(operand);
+        }
+      }
+    }
+
+    const auto newNumOperandPairs = usedResults.size();
+    const auto newNumOperands = newNumOperandPairs * pairSize;
+    if (newNumOperands != usedArgs.count())
+      return rewriter.notifyMatchFailure(op, [&](Diagnostic &diag) {
+        diag << "non-conservative case: expected " << usedArgs.count()
+             << " arguments to match " << newNumOperandPairs
+             << " return operands, but got " << newNumOperands;
+      });
```

Comment:
Yes, you are right, it can be dropped safely. But i am not sure that it is worth doing partial pruning, without deciding what should be done with `r2`. Ideally it should be transformed into:
```
//  (a, a1) (c, c1)
//    ▴       ▴
//    ├───────┘
//    │
//    │
//    │
//    r0      #0
//    U
```

> My understanding is that with this check we only consider cases where each used result uses only one operand pair.

i'd say where number of used results matches number of used operand pairs, e.g. this works:
```
//  (a, a1) (b, b1) (c, c1)
//    ▴       ▴       ▴
//    │       │       │
//    ├───────┼───────┘
//    │◄──────┼───────┐
//    │       │       │
//    r0      r1      r2
//    U               U
```
r0 uses (a, a1) (c, c1)
r1 uses (b, b1)
r2 used (a, a1) (c, c1)

Reply:
Gotcha, that all makes sense to me, and I think this is fine. Consider adding cases with 2 used results and even all 3 results used (just to be safe).

Reply:
I have added tests and suppose that there are actually 3 cases in total:
```
          1              //            2              //            3
                         //                           //
(a, a1) (b, b1) ( ... )  //  (a, a1) (b, b1) ( ... )  //  (a, a1) (b, b1) ( ... )
  ▴                      //    ▴       ▴              //    ▴       ▴
  │                      //    │       │              //    │       │
  │◄──────┐       ...    //    ├───────┘       ...    //    │       │       ...
  │       │              //    │                      //    │       │
  │       │              //    │                      //    │       │
  r0      r2      ...    //    r0      r2      ...    //    r0      r1      ...
  U       U              //    U                      //    U
                         //                           //
```
case 1: |{used results}| >  |{used operand pairs}|
case 2: |{used results}| <  |{used operand pairs}|
case 3: |{used results}| == |{used operand pairs}|

Currently, we only support case 3. I was thinking, if we could somehow reduce cases 1 and 2 to case 3, could we benefit from automatic pruning like we do in case 3? For case 1, it seems simple - just select any unused operand pair(s) to match the number of used results. For case 2, we can either succeed or fail to match the number of used operand pairs. To succeed, we need to find an unused result where the `{ reachable block arguments }` is a subset of `{ set of used block arguments }`. If we fail, we might not get automatic pruning, but we could still do it manually. However, a manual approach would likely require the same level of analysis. Perhaps, instead of using `subset`, we could consider using `intersect` and gradually work towards creating a pruned version of the operation. Let's revisit the extended `reduce_unused_case6` that we previously discussed.

```
(a, a1) (b, b1) (c, c1) (d, d1)
  ▴       ▴       ▴       ▴
  │       ├──────►│       │
  │       │◄──────│       │
  │◄──────┼───────┤       │
  │       │       │       │
  r0      r1      r2      r3
          U
```
We might start from r1, see that it requires two operand pairs (b, b1), (c, c1). That would require 2 return results. Lets find a candidate that at least intersects with those 2 operand pairs. Lets say it is r2, that additionally uses (a, a1) pair, therefore we are forced to include that too. (a, a1) is used by r0, include that too. Add everything to sets. Repeat until |{~~used~~ results}| == |{~~used~~ operand pairs}|. Finally we should stop at size 3. r3 plus (d, d1) would be pruned.

Honestly i don't like such complicated approach for this problem. I think doing strict `subset` should be enough, everything else is unsupported. We can probably leave current solution as it is, and i might work on case1 -> case3 and case2 -> case3 (subset version) later.

Reply:
With this newfound [understanding](https://github.com/openxla/stablehlo/pull/2165#discussion_r1558159623), the entire writeup may be partially or completely incorrect. I will write it down here so it does not mislead anyone, including myself, if someone wishes to further improve this.